tests/test_config.py

```diff
@@ -0,0 +1,330 @@
+import sys
+from contextlib import nullcontext as does_not_raise
+from inspect import signature
+from typing import Any, ContextManager, Iterable, NamedTuple, Type, Union
+
+from dirty_equals import HasRepr
+
+from pydantic import BaseConfig, BaseModel, Extra, Field, PrivateAttr, PydanticSchemaGenerationError, ValidationError
+from pydantic.config import ConfigDict, _default_config
+from pydantic.errors import PydanticUserError
+
+if sys.version_info < (3, 9):
+    from typing_extensions import Annotated
+else:
+    from typing import Annotated
+
+import pytest
+
+
+@pytest.fixture(scope='session', name='BaseConfigModelWithStrictConfig')
+def model_with_strict_config():
+    class ModelWithStrictConfig(BaseModel):
+        a: int
+        # strict=False overrides the Config
+        b: Annotated[int, Field(strict=False)]
+        # strict=None or not including it is equivalent
+        # lets this field be overridden by the Config
+        c: Annotated[int, Field(strict=None)]
+        d: Annotated[int, Field()]
+
+        class Config:
+            strict = True
+
+    return ModelWithStrictConfig
+
+
+def _equals(a: Union[str, Iterable[str]], b: Union[str, Iterable[str]]) -> bool:
+    """
+    Compare strings with spaces removed
+    """
+    if isinstance(a, str) and isinstance(b, str):
+        return a.replace(' ', '') == b.replace(' ', '')
+    elif isinstance(a, Iterable) and isinstance(b, Iterable):
+        return all(_equals(a_, b_) for a_, b_ in zip(a, b))
+    else:
+        raise TypeError(f'arguments must be both strings or both lists, not {type(a)}, {type(b)}')
+
+
+@pytest.mark.filterwarnings('ignore:.* is deprecated.*:DeprecationWarning')
+class TestsBaseConfig:
+    def test_base_config_equality_defaults_of_config_dict_class(self):
+        for key, value in _default_config.items():
+            assert getattr(BaseConfig, key) == value
```

Comment:
@samuelcolvin prior to my changes, there was an additional line in this test:
```python
assert _default_config.keys() == {k for k in dir(BaseConfig) if not k.startswith('__')}
```
I'm assuming it's okay that `dir(BaseConfig)` is now empty; I noticed that `dir(Config)` is used in the non-test code, but the relevant code seems to still work properly with some properties just not defined (instead of always defined with the default value).

Considering all the tests are passing I think it should be okay, I just want to make sure you aren't aware of anything I might be overlooking.

Reply:
yes, that's the idea.

We should just add one explicit test that `BaseConfig.something` returns what we expect (and yields a warning).

I guess we should also check that `BaseConfig().something` does the same, as I'm sure somewhere someone is initialising it.